core-rust/state-manager/src/protocol/protocol_config.rs

```diff
@@ -0,0 +1,178 @@
+use radix_engine::prelude::{dec, ScryptoCategorize, ScryptoDecode, ScryptoEncode};
+use radix_engine_common::math::Decimal;
+use radix_engine_common::prelude::{hash, scrypto_encode};
+
+use radix_engine_common::types::Epoch;
+
+use crate::ProtocolUpdaterFactory;
+use utils::btreeset;
+
+// This file contains types for node's local static protocol configuration
+
+pub const GENESIS_PROTOCOL_VERSION: &str = "babylon-genesis";
+pub const ANEMONE_PROTOCOL_VERSION: &str = "anemone";
+
+const MAX_PROTOCOL_VERSION_NAME_LEN: usize = 16;
+
+/// Returns a protocol version name left padded to canonical length (16 bytes)
+pub fn padded_protocol_version_name(unpadded_protocol_version_name: &str) -> String {
+    let mut res = "".to_owned();
+    for _ in 0..16 - unpadded_protocol_version_name.len() {
+        res.push('0');
+    }
+    res.push_str(unpadded_protocol_version_name);
+    res
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
+pub struct ProtocolUpdate {
+    pub next_protocol_version: String,
+    pub enactment_condition: ProtocolUpdateEnactmentCondition,
+}
+
+impl ProtocolUpdate {
+    pub fn readiness_signal_name(&self) -> String {
+        // Readiness signal name is 32 bytes long and consists of:
+        // - 16 hexadecimal chars of leading bytes of `hash(enactment_condition + next_protocol_version)`
+        // - next_protocol_version: 16 bytes,
+        //      left padded with ASCII 0's if protocol version name is shorter than 16 characters
+        let mut bytes_to_hash = scrypto_encode(&self.enactment_condition).unwrap();
+        bytes_to_hash.extend_from_slice(self.next_protocol_version.as_bytes());
+        let protocol_update_hash = hash(&bytes_to_hash);
+        let mut res = hex::encode(protocol_update_hash)[0..16].to_string();
+        res.push_str(&padded_protocol_version_name(&self.next_protocol_version));
+        res
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
+pub struct ProtocolConfig {
+    pub genesis_protocol_version: String,
+    pub protocol_updates: Vec<ProtocolUpdate>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
+pub enum ProtocolUpdateEnactmentCondition {
+    EnactWhenSupportedAndWithinBounds {
+        lower_bound: Epoch,
+        upper_bound: Epoch,
```

Comment:
> This might be a little clearer as EnactAtStartOfAnEpochIfConditionsMet ?

I personally like it less specific, it leaves the doors open for "at state version" updates (by extending the bounds to `enum ProtocolUpdateBound { Epoch(..), StateVersion(..) }`.

I'll add a comment :+1: both are actually inclusive

Reply:
> both are actually inclusive

@LukasGasior1 just to avoid confusion, I'd vote to do the classic "from inclusive, to exclusive". The use of `Range<>` would achieve it while boosting the OOP score even further :p

Alternatively, if it conceptually makes LOTS of sense to operate on "to inclusive", then the best names would be `min, max` (since "max" hints "maximum possible" or "maximum that is still allowed", and won't be confused with "exclusive").

Reply:
> I personally like it less specific, it leaves the doors open for "at state version" updates (by extending the bounds to enum ProtocolUpdateBound { Epoch(..), StateVersion(..) }.

I disagree - when the enactment is checked is key to the condition. Having it enact only at a start of an epoch is different to having it enact at any particular state version.

I assume this condition can only be used to enact at start of epoch, right? Like we can't have this thing trigger mid-epoch?

Reply:
> Like we can't have this thing trigger mid-epoch?

I personally hope we will never have the "trigger protocol update mid-epoch" feature. I'd prefer to have a way of forcing a premature epoch change (with the Engine's cooperation, which Josh confirmed seems possible).

Reply:
>  assume this condition can only be used to enact at start of epoch, right? Like we can't have this thing trigger mid-epoch?

Currently yes, but with state version updates the same "enact when supported" condition could be reused. One part of the condition would be met at the beginning of an epoch (the support sub-condition) and the other part possibly mid-epoch (min/max bound sub-condition).

Anyway, happy to rename that for now and revisit if/when we actually introduce other conditions.

Reply:
but just for the sake of discussion :smile: 

> I disagree - when the enactment is checked is key to the condition. Having it enact only at a start of an epoch is different to having it enact at any particular state version.

I don't think when the enactment is checked is key. I treat the condition as being continously evaluated: if at *any* time (i.e. after any txn) all its components are met, then it passes. It just so happens that currently the state that influences the evaluation changes only at epoch boundaries

Reply:
(as discussed, will rename in a follow up PR)